### PR TITLE
Add ability to EBS Surrogate builder to persist properties from Marketplace source image, such as product codes

### LINKED ISSUE #440: Need an improved chroot-builder
#### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

Please search the existing issues for relevant feature requests, and use the
reaction feature
(https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)
to add upvotes to pre-existing requests.

#### Description

Need the ability to have a chroot-build method that supports a workflow that looks like:
1. Launch build-EC2 with root and build-target EBS attached
2. Execute a build to the secondary EBS
3. Power off the build-EC2
4. Detach boot and build-target EBS from build-EC2
5. Re-attach build-target EBS as build-EC2's (new) boot-EBS
6. Register stopped build-EC2 as image
7. Terminate build-EC2 
8. Clean up original boot-EBS and any other resources created for build

#### Use Case(s)

In order to inherit AMI/EC2 attributes from the bootstrap AMI &ndash; primarily concerned with the `billingProduct` &ndash; it's necessary to have this kind of EBS-switching workflow. The current amazon builder options don't afford this. 

Previously, we used the amazon-ebs builder combined with a ["pivot-root"](https://github.com/plus3it/spel/blob/master/spel/scripts/pivot-root.sh) preparation-script and a [disk-releaser stage](https://github.com/plus3it/spel/blob/master/spel/minimal-linux.pkr.hcl#L918-L925) in our workflow. This allowed simply re-writing the contents of the boot-disk.

Unfortunately, it appears that there's a bug that got introduced in the Linux 5.x kernel (and that also seems to exist forward into at least the 6.1 kernel) that prevents the kernel from recognizing live/on-the-fly disk-layout changes made to the boot-disk. While issues have been opened with Red Hat (impacts their 5.14-based RHEL 9.x releases) and AWS (impacts their 6.1-based Amazon 2023 images) and each vendor has reproduced the problem, no determination of if or when a fix has been given by either vendor.

Ultimately, using the amazon-ebs builder the way we did was kind of hackish, but, "it worked". Prior to the recently discovered issues when attempting to extend our automation to RHEL 9.x and Amazon 2023 AMIs, it never necessitated looking at other Amazon Packer-builders. The current amazon-chroot looks _close_ to what we need, but seems to lack the ability do anything but steps 1 & 2 (as described above). Steps 3+ are the key to producing images that inherit attributes from the bootstrap AMI. If we had an improved chroot-style build, we'd probably move all of our AMI-build recipies to that builder-option.

#### Potential configuration

```
```

#### Potential References

* [Inheriting billingProduct attribute](https://thjones2.blogspot.com/2015/03/so-you-dont-want-to-byol.html)
* Red Hat Issue ID [RHEL-17421](https://issues.redhat.com/browse/RHEL-17421)

I would link to the Amazon issue ID, however, it is not subscribable outside of the account it was opened from

### PR DESCRIPTION
**Pull request by dwc0011**
Add option to swap the root device to an EBS surrogate and create a new AMI in addition to being able to register an AMI.  This allows billing data to be retained when creating AMI's.  

Closes #440

### CODE REVIEW

**Comment by dwc0011** [comment]
Wanted to bump this to see if we can someone has time to review this.  Thanks

**Comment by lorengordon** [comment]
@nywilken Sorry for the ping, but just wanted to check and see if we could get a review on this pr, or an idea whether we're on an acceptable track. What I really need to know is if there's no possibility of merging this, so we can refocus on trying to find some other way to manage the workflow. Thanks!

**Comment by nywilken** [comment]
> @nywilken Sorry for the ping, but just wanted to check and see if we could get a review on this pr, or an idea whether we're on an acceptable track. What I really need to know is if there's no possibility of merging this, so we can refocus on trying to find some other way to manage the workflow. Thanks!

@lorengordon @dwc0011 apologies I've been out sick for the last few weeks and circling back to this PR. We have plans to review and merge.

File: builder/ebssurrogate/step_swap_volumes.go
Comment (nywilken) [inline builder/ebssurrogate/step_swap_volumes.go:18]: This looks to be the comment from SnapshotVolumes

```suggestion
// StepSnapshotVolumes creates snapshots of the created volumes.
```

File: builder/ebssurrogate/builder.go
Comment (nywilken) [inline builder/ebssurrogate/builder.go:322]: Given the how far down the use of these steps are I find the name to be too generic. Let me think about this a little but I do wonder if we should either use a `multistep.If(b.config.RootDevice.ImageMethod == "create", StepCreateAMI{...})` or if we should just add both the register and create steps to the steps slice and in each step check the image method to determine if the step should run.

**Review by nywilken — COMMENTED** [review]
I took a first pass and left a few comments. I think the if statement is simple enough but I would like for the names to be a bit more explicit on the work that is being done. ImageMethod and VolumeMethod seem simple. 

If left a question around the DeRegister step being needed. Please let me know.

File: builder/ebssurrogate/builder.go
Comment (dwc0011) [reply builder/ebssurrogate/builder.go:437]: I would say yes since the create image also register the image for you.  See below
 
"Note
For Amazon EBS-backed instances, CreateImage creates and registers the AMI in a single request, so you don't have to register the AMI yourself. We recommend that you always use CreateImage unless you have a specific reason to use RegisterImage."

https://docs.aws.amazon.com/cli/latest/reference/ec2/register-image.html#:~:text=For%20Amazon%20EBS%2Dbacked%20instances,an%20AMI%20at%20any%20time.

File: builder/ebssurrogate/builder.go
Comment (dwc0011) [reply builder/ebssurrogate/builder.go:322]: Let me know which route you think would be best and I will refactor it.  I attempted to model this on the existing conditional instance step.

File: builder/ebssurrogate/builder.go
Comment (lorengordon) [reply builder/ebssurrogate/builder.go:437]: As confirmation, you can also see that the Deregister step is part of the `ebs` builder, which also uses CreateImage.

https://github.com/hashicorp/packer-plugin-amazon/blob/7710fbc3554e464faca17de05dd9e820e062a829/builder/ebs/builder.go#L409-L415

**Comment by nywilken** [comment]
Hi @dwc0011 @lorengordon thanks again for pushing this change forward and for you patience with the review. Speaking about this change internally we found ourselves a bit confused on the approach. To be clear, we understand why and what is happening. But we do wonder if there is a different approach on using create or register that we can take to make this clearer to not confuse users who may already be relying on ebssurrogate. 

To help push this forward @lbajolet-hashicorp and I would like to setup a call so we can chat a bit more. Would you be up for a video call this week? We are both based in EST timezone. 

If you are up for it you can send an email to <packer at hashicorp dot com>

**Comment by lorengordon** [comment]
Sure, @nywilken, we can hop on a call. I'm on Pacific Time, @dwc0011 is Eastern Time. Though, it looks like the email address was cutoff? If it's easier, I'm in the hangops slack as well, as `@loren`.

**Comment by lbajolet-hashicorp** [comment]
Hi @lorengordon,

Thanks for the quick update; the email address was cutoff since Markdown interpreted it as raw html because of the braces, I've encased it in code markup, you should be able to see it now.

**Comment by dwc0011** [comment]
I am definitely open to a call,  I am not available from 0800-1200 on Wednesday and I am out of office on Friday.  Other than those times I am available!

Thanks,
Dennis
________________________________

**Comment by lorengordon** [comment]
> Hi @lorengordon,
> 
> Thanks for the quick update; the email address was cutoff since Markdown interpreted it as raw html because of the braces, I've encased it in code markup, you should be able to see it now.

Roger that, email sent!

**Comment by lorengordon** [comment]
@nywilken @lbajolet-hashicorp Per the offline discussion, Dennis renamed the new argument as a bool, `use_create_image` and updated the argument description. Please let us know if you need anything else!

File: builder/ebssurrogate/step_swap_volumes.go
Comment (lbajolet-hashicorp) [inline builder/ebssurrogate/step_swap_volumes.go:108]: I would suggest renaming the function here to detachVolume, mostly since camelCase is the norm in Go
```suggestion
func (s *StepSwapVolumes) detachVolume(ctx context.Context, ec2conn *ec2.EC2, deviceName string, volumeId string) error {
```

**Review by lbajolet-hashicorp — APPROVED** [review]
Hi @dwc0011, @lorengordon 

Looks good to me! I left a mini nitpick on a function's name, but aside from that, we should be good to go very soon.
I'll let @nywilken do a quick final review in addition to mine before we merge, and we'll follow up with a release likely next week.

Thanks for the reroll and the chat earlier, good to see this feature in the plugin, really, thanks for your efforts!

**Comment by dwc0011** [comment]
Thank you for your time today, look forward to this being incorporated into packer!

File: builder/ebssurrogate/builder.go
Comment (nywilken) [inline builder/ebssurrogate/builder.go:321]: 💯

File: builder/ebssurrogate/builder.go
Comment (nywilken) [inline builder/ebssurrogate/builder.go:324]: By moving this attribute up one level this should become 
```go
	if b.config.UseCreateImage {
```

**Review by nywilken — CHANGES_REQUESTED** [review]
Thanks again for the syncing and re-rolling this change. Overall it looks good. I left a couple of suggested changes to have the configuration attribute on the builder config and not within the ami_root_device config. 

I'm running the acceptance test now and will report back if I run into any issues.

**Comment by lorengordon** [comment]
> Thanks again for the syncing and re-rolling this change. Overall it looks good. I left a couple of suggested changes to have the configuration attribute on the builder config and not within the ami_root_device config.
> 
> I'm running the acceptance test now and will report back if I run into any issues.

Dennis is out today, but will tackle the requested change next week. Or feel free to make the changes yourself if you want to close it sooner.

**Comment by dwc0011** [comment]
@nywilken I have addressed all the moves and fixed the tests as well.  I appreciate the feedback and pointing us in the right direction.  I am new to go and packer in general and expected change requests, hopefully they are all complete and we are set!

Thank you again!

**Review by nywilken — APPROVED** [review]
@dwc0011 thanks for making the updated changes and pushing this change forward. This all looks good and the added acceptance tests are all green. Cheers! 

```=== RUN   TestAccBuilder_EbssurrogateUseCreateImageTrue
2024/03/04 11:59:30 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
2024/03/04 11:59:30 Found region us-west-2
2024/03/04 11:59:30 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
--- PASS: TestAccBuilder_EbssurrogateUseCreateImageTrue (198.57s)
=== RUN   TestAccBuilder_EbssurrogateUseCreateImageFalse
2024/03/04 12:01:32 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
2024/03/04 12:01:32 Found region us-west-2
2024/03/04 12:01:32 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
--- PASS: TestAccBuilder_EbssurrogateUseCreateImageFalse (122.11s)
=== RUN   TestAccBuilder_EbssurrogateUseCreateImageOptional
2024/03/04 12:03:03 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
2024/03/04 12:03:03 Found region us-west-2
2024/03/04 12:03:03 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
--- PASS: TestAccBuilder_EbssurrogateUseCreateImageOptional (91.61s)
PASS
ok      github.com/hashicorp/packer-plugin-amazon/builder/ebssurrogate  413.330s
```

**Comment by nywilken** [comment]
We release every Tuesday so this will go out tomorrow. If I can get it released today I will trigger things earlier. Thanks again for making this change and for working with @lbajolet-hashicorp and I to better understand the needed.

**Pull request merged**
